File: projects/IST-IG-DD-JAVA/transforms-java/src/main/java/myproject/datasets/exposure/typed/fp_gas_m0145.java
package myproject.datasets.exposure.typed;

import com.palantir.transforms.lang.java.api.Compute;
import com.palantir.transforms.lang.java.api.Input;
import com.palantir.transforms.lang.java.api.Output;
import com.palantir.transforms.lang.java.api.FoundryOutput;
import com.palantir.transforms.lang.java.api.FoundryInput;
import com.palantir.transforms.lang.java.api.ReadRange;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.functions;
import myproject.datasets.util.BP_ReadRangeUtils;
import myproject.datasets.util.BP_WriteRangeUtils;

/**
 * This is an example high-level Transform intended for automatic registration.
 */
public final class fp_gas_m0145 {

    // The class for an automatically registered Transform contains the compute
    // function and information about the input/output datasets.
    // Automatic registration requires "@Input" and "@Output" annotations.
    @Compute    
    public void myComputeFunction(@Input("/BP/IST-IG-SS-Systems/data/raw/freeport/fp_gas_m0145/fp_gas_m0145") FoundryInput input,
      @Output("/BP/IST-IG-DD/data/technical/exposure_java/typed/freeport/fp_gas_m0145") FoundryOutput output
    ) {
        ReadRange readRange = BP_ReadRangeUtils.getReadRange(input);
        Dataset<Row> df = input.asDataFrame().readForRange(readRange);
        df = df.select(
                df.col("Request_ID").cast("int"),
                functions.to_date(df.col("Analysis_Date"), "dd/MM/yyyy").alias("Analysis_Date"),
                df.col("Request_Type"),
                df.col("Contract_Month_DMD").cast("int"),
                df.col("Deal_Tracking_Num").cast("int"),
                df.col("Business_Unit"),
                df.col("Trader"),
                df.col("Portfolio"),
                df.col("External_BU"),
                df.col("Internal_BU"),
                df.col("Currency"),
                df.col("Instrument_Type"),
                functions.to_date(df.col("Trade_Date"), "dd/MM/yyyy").alias("Trade_Date"),
                functions.to_date(df.col("Trade_Time"), "dd/MM/yyyy").alias("Trade_Time"),
                df.col("Toolset"),
                df.col("External_Portfolio"),
                df.col("Broker"),
                df.col("OTC_Clearing_Broker").cast("int"),
                df.col("Offset_Transaction_Number").cast("int"),
                df.col("Reference"),
                df.col("Projection_Index"),
                df.col("UOM"),
                df.col("Grid_Point_ID").cast("int"),
                df.col("Grid_Point_Name"),
                functions.to_date(df.col("Grid_Point_Start_Date"), "dd/MM/yyyy").alias("Grid_Point_Start_Date"),
                functions.to_date(df.col("Grid_Point_End_Date"), "dd/MM/yyyy").alias("Grid_Point_End_Date"),
                functions.to_date(df.col("Grid_Point_Start_Date_Long"), "dd/MM/yyyy").alias("Grid_Point_Start_Date_Long"),
                functions.to_date(df.col("Grid_Point_End_Date_Long"), "dd/MM/yyyy").alias("Grid_Point_End_Date_Long"),
                df.col("Delta").cast("double"),
                df.col("Gamma").cast("double"),
                df.col("Vega").cast("double"),
                df.col("Theta").cast("double"),
                df.col("Risk_Factor_Name"),
                df.col("Delta_Shift").cast("double"),
                df.col("Gamma_Factor").cast("int"),
                df.col("Risk_Factor_CCY"),
                df.col("CCY_Conversion_Rate").cast("double"),
                df.col("Risk_Factor_UOM"),
                df.col("UOM_Conversion_Rate").cast("double"),
                df.col("MR_Delta").cast("double"),
                df.col("MR_Gamma").cast("double"),
                df.col("MR_Vega").cast("double"),
                df.col("MR_Theta").cast("double"),
                df.col("Market_Value").cast("double"),
                df.col("Bucket_Name"),
                df.col("Bucket_Id").cast("int"),
                df.col("Bucket_Order").cast("int"),
                functions.to_date(df.col("Bucket_Start_Date"), "dd/MM/yyyy").alias("Bucket_Start_Date"),
                functions.to_date(df.col("Bucket_End_Date"), "dd/MM/yyyy").alias("Bucket_End_Date"),
                df.col("Extract_Type"),
                df.col("Type"),
                functions.to_timestamp(df.col("Run_Datetime"), "yyyy-MM-dd HH:mm:ss").alias("Run_Datetime")
                );
        output.getDataFrameWriter(df).write(BP_WriteRangeUtils.getWriteRange(input));
    }
}

File: latest_integration/transforms-python/src/python/exposure/typed/fp_gas_m0145.py
from transforms.api import transform_df, incremental, Input, Output
import pyspark.sql.functions as F


@incremental()
@transform_df(
    Output("/BP/IST-IG-DD/data/technical/exposure/typed/freeport/fp_gas_m0145"),
    df=Input("/BP/IST-IG-SS-Systems/data/raw/freeport/fp_gas_m0145/fp_gas_m0145"),
)
def my_compute_function(df):
    """
    Transformation stage function
    Version: T2
    This function:
        For processing raw data and converting to type
        Reorders the dataframe using select statement.
    Args:
        df (dataframe): current fp_gas_145m

    Returns:
        dataframe: transformed fp_gas_145m
    """
    df = df.select(
        df['Request_ID'].cast('int'),
        F.to_date(df['Analysis_Date'], 'dd/MM/yyyy').alias('Analysis_Date'),
        df['Request_Type'],
        df['Contract_Month_DMD'].cast('int'),
        df['Deal_Tracking_Num'].cast('int'),
        df['Business_Unit'],
        df['Trader'],
        df['Portfolio'],
        df['External_BU'],
        df['Internal_BU'],
        df['Currency'],
        df['Instrument_Type'],
        F.to_date(df['Trade_Date'], 'dd/MM/yyyy').alias('Trade_Date'),
        F.to_date(df['Trade_Time'], 'dd/MM/yyyy').alias('Trade_Time'),
        df['Toolset'],
        df['External_Portfolio'],
        df['Broker'],
        df['OTC_Clearing_Broker'].cast('int'),
        df['Offset_Transaction_Number'].cast('int'),
        df['Reference'],
        df['Projection_Index'],
        df['UOM'],
        df['Grid_Point_ID'].cast('int'),
        df['Grid_Point_Name'],
        F.to_date(df['Grid_Point_Start_Date'], 'dd/MM/yyyy').alias('Grid_Point_Start_Date'),
        F.to_date(df['Grid_Point_End_Date'], 'dd/MM/yyyy').alias('Grid_Point_End_Date'),
        F.to_date(df['Grid_Point_Start_Date_Long'], 'dd/MM/yyyy').alias('Grid_Point_Start_Date_Long'),
        F.to_date(df['Grid_Point_End_Date_Long'], 'dd/MM/yyyy').alias('Grid_Point_End_Date_Long'),
        df['Delta'].cast('double'),
        df['Gamma'].cast('double'),
        df['Vega'].cast('double'),
        df['Theta'].cast('double'),
        df['Risk_Factor_Name'],
        df['Delta_Shift'].cast('double'),
        df['Gamma_Factor'].cast('int'),
        df['Risk_Factor_CCY'],
        df['CCY_Conversion_Rate'].cast('double'),
        df['Risk_Factor_UOM'],
        df['UOM_Conversion_Rate'].cast('double'),
        df['MR_Delta'].cast('double'),
        df['MR_Gamma'].cast('double'),
        df['MR_Vega'].cast('double'),
        df['MR_Theta'].cast('double'),
        df['Market_Value'].cast('double'),
        df['Bucket_Name'],
        df['Bucket_Id'].cast('int'),
        df['Bucket_Order'].cast('int'),
        F.to_date(df['Bucket_Start_Date'], 'dd/MM/yyyy').alias('Bucket_Start_Date'),
        F.to_date(df['Bucket_End_Date'], 'dd/MM/yyyy').alias('Bucket_End_Date'),
        df['Extract_Type'],
        df['Type'],
        F.to_timestamp(df['Run_Datetime'], 'yyyy-MM-dd HH:mm:ss').alias('Run_Datetime'),
    )
    return df
